satellite/satellitedb/repairqueue.go

```diff
@@ -21,39 +22,36 @@ type repairQueue struct {
 	db *dbx.DB
 }
 
-func (r *repairQueue) Enqueue(ctx context.Context, seg *pb.InjuredSegment) error {
-	val, err := proto.Marshal(seg)
-	if err != nil {
-		return err
+func (r *repairQueue) Insert(ctx context.Context, seg *pb.InjuredSegment) error {
+	_, err := r.db.ExecContext(ctx, r.db.Rebind(`INSERT INTO injuredsegments ( path, data ) VALUES ( ?, ? )`), seg.Path, seg)
+	if err != nil && (strings.Contains(err.Error(), "UNIQUE constraint failed") || strings.Contains(err.Error(), "violates unique constraint")) {
+		return nil // quietly fail on reinsert
 	}
-
-	_, err = r.db.Create_Injuredsegment(
-		ctx,
-		dbx.Injuredsegment_Info(val),
-	)
 	return err
 }
 
-func (r *repairQueue) postgresDequeue(ctx context.Context) (seg pb.InjuredSegment, err error) {
-	err = r.db.DB.QueryRowContext(ctx, `
-	DELETE FROM injuredsegments
-		WHERE id = ( SELECT id FROM injuredsegments ORDER BY id FOR UPDATE SKIP LOCKED LIMIT 1 )
-		RETURNING info
-	`).Scan(&seg)
+func (r *repairQueue) postgresSelect(ctx context.Context) (seg *pb.InjuredSegment, err error) {
+	//todo :  add or age > some time (1 hour?)
+	err = r.db.QueryRowContext(ctx, r.db.Rebind(`
+	UPDATE injuredsegments SET attempted = ? WHERE path = (
+		SELECT path FROM injuredsegments
+		WHERE attempted IS NULL
+		ORDER BY path FOR UPDATE SKIP LOCKED LIMIT 1
+	) RETURNING data`), time.Now().UTC()).Scan(&seg)
 	if err == sql.ErrNoRows {
 		err = storage.ErrEmptyQueue.New("")
 	}
-	return seg, err
+	return
 }
 
-func (r *repairQueue) sqliteDequeue(ctx context.Context) (seg pb.InjuredSegment, err error) {
+func (r *repairQueue) sqliteSelect(ctx context.Context) (seg *pb.InjuredSegment, err error) {
 	err = r.db.WithTx(ctx, func(ctx context.Context, tx *dbx.Tx) error {
-		var id int64
-		err = tx.Tx.QueryRowContext(ctx, `SELECT id, info FROM injuredsegments ORDER BY id LIMIT 1`).Scan(&id, &seg)
+		var path string
+		err = tx.Tx.QueryRowContext(ctx, r.db.Rebind(`SELECT path, data FROM injuredsegments WHERE attempted IS NULL ORDER BY path LIMIT 1`)).Scan(&path, &seg)
 		if err != nil {
 			return err
```

Comment:
Shouldn't we treat `sql.ErrRows` different here?

Reply:
do you mean sql.ErrNoRows?

Reply:
because that line is part of anonymous function that get run via `r.db.WithTx(ctx, func(ctx context.Context, tx *dbx.Tx) error) `, it is sufficient to return any error and handle ErrNoRows in the main function.